RandLAPACK/comps/rl_preconditioners.hh

```diff
@@ -217,5 +219,101 @@ int64_t make_right_orthogonalizer(
     return rank;
 }
 
+/**
+ * Construct data (V, eigvals) for a Nystrom preconditioner of regularized linear
+ * systems involving a PSD matrix A.
+ * 
+ *      If our regularized linear system is 
+ *          (A + mu*I)x= b,
+ *      then the preconditioner would be
+ *          P^{-1} = V diag(eigvals + mu)^{-1} V' + (I - VV') / (min(eigvals) + mu).
+ * 
+ *      Such a preconditioners will lead to (A + mu*I)*P^{-1} being well-conditioned
+ *      if the spectral norm error || A - V diag(eigvals) V' || is no larger than mu.
+ * 
+ * This function computes (V, eigvals) in an iterative process. The matrix V has "k_in"
+ * columns at the first iteration, where "k_in" the value of k on entry. Each iteration
+ * ends by computing an estimate for || A - V diag(eigvals) V' ||. If this estimate
+ * falls below tol = mu_min/5, then we return. Otherwise, we double k and try again.
+ * 
+ * Optional arguments to this function can trade-off greater computational cost
+ * with (1) better approximations A_hat = V diag(eigvals) V' for fixed k and (2)
+ * better estimates of the spectral norm error ||A - A_hat||.
+ * 
+ * @param[in] uplo
+ *      blas::Uplo::Lower or blas::Uplo::Upper. Specifies the triangular
+ *      part of A that we access while computing (V, eigvals).
+ * @param[in] A
+ *      A buffer for an m \times m positive semidefinite matrix A.
```

Comment:
What's this about Gram matrices? Every PSD matrix is indeed a Gram matrix of some other matrix, but that doesn't show up here. Do you mean to ask if ``A`` is an explicit array, rather than a positive semidefinite linear operator?

Reply:
Ah sorry yeah I meant that we're requiring `A` be an explicit PSD matrix; but that seems really disappointing as someone making a least squares solver with this who would prefer to never construct a Gram matrix.

Reply:
@PTNobel I've fixed this with a very simple linear operator interface. The size of the PR expanded but you can skip over many of the changed files.